File: 2336-smallest-number-in-infinite-set/2336-smallest-number-in-infinite-set.cpp
class SmallestInfiniteSet {
public:
    priority_queue<int, vector<int>, greater<int>> heap;
    unordered_set<int> exist;
    
    SmallestInfiniteSet() {
        for (int i=0; i<1000; i++)
        {
            heap.push(i+1);
            exist.insert(i+1);
        }
    }
    
    int popSmallest() {
        int tmp = heap.top();
        heap.pop();
        exist.erase(tmp);
        return tmp;
    }
    
    void addBack(int num) {
        if (!exist.contains(num))
        {
            heap.push(num);
            exist.insert(num);
        }
    }
};

/**
 * Your SmallestInfiniteSet object will be instantiated and called as such:
 * SmallestInfiniteSet* obj = new SmallestInfiniteSet();
 * int param_1 = obj->popSmallest();
 * obj->addBack(num);
 */

File: 2336-smallest-number-in-infinite-set/2336-smallest-number-in-infinite-set.py
class SmallestInfiniteSet:

    def __init__(self):
        self.heap = list(range(1,1001))
        self.set = set(self.heap)
        heapq.heapify(self.heap)

    def popSmallest(self) -> int:
        tmp = heapq.heappop(self.heap)
        self.set.remove(tmp)
        return tmp

    def addBack(self, num: int) -> None:
        if num not in self.set:
            heapq.heappush(self.heap, num)
            self.set.add(num)


# Your SmallestInfiniteSet object will be instantiated and called as such:
# obj = SmallestInfiniteSet()
# param_1 = obj.popSmallest()
# obj.addBack(num)